Model/ReceiptDataProvider.php

```diff
@@ -332,7 +332,7 @@ protected function processOrder($paymentVerified)
             $this->currentOrder->setState(Recurring::ORDER_STATE_CUSTOM_CODE);
             $this->currentOrder->setStatus(Recurring::ORDER_STATUS_CUSTOM_CODE);
             $this->currentOrder->addCommentToStatusHistory(__('Pending payment from Paytrail Payment Service'));
-        } else {
+        } elseif (!($this->currentOrder->getStatus() === 'processing' && $this->currentOrder->getTotalPaid())) {
```

Comment:
As I sometimes get confused with these negative if statements. So deal here is that if status is something else than processing OR total paid is 0 we see payment as complete OR if both checks are seen as false we do the same.

So if you can with few words explain the logic between this approach it would be helpful for me also to understand.

Reply:
Basically it validate that FAILED payment for already paid/processing order doesn't cancel this order.
Before was like, when already paid order get FAILED payment, then order was canceled and we can't do anything with this canceled order. :)
So it means that if order IS NOT 'processing' & order doesn't have TotalPaid value, then you can cancel this order.
But we are working to change this cancelation, put order on hold or something.